# Final Project: Black Friday Excel Data

In [1]:
import openpyxl as xl
import tax_calculator as tc
import numpy as np
# pprint prints dictionaries a bit more nicely than print
from pprint import pprint

## Part 1

First, read in maven_ski_shop_data.xlsx

Then, assign the orders worksheet to an intuitive variable name. 

In [2]:
wb = xl.load_workbook('maven_ski_shop_data.xlsx')
wb.sheetnames
orders = wb['Orders_Info']

### Column Printer Function

Since we'll be working with Excel frequently, let's create a helper function to print all the rows a specified column.

This will help us view imported data without looking at Excel, and review the data we've written into columns without needing to save a file to review the changes.

Your function should take two arguments:

* Worksheet where data is located
* Column Letter of data to print

Your function should:

* Print Cell Coordinate (e.g. A1, D4).
* Print the contents of the cell.

Then call your function twice:

* Once to print Order ID (column A in 'Orders_Info')
* Once to print subtotal (column D in 'Orders_Info')

In [3]:
def column_printer(worksheet, column_to_print):
    for i in range(1, worksheet.max_row + 1):
        print(f'{column_to_print}{i}', worksheet[f'{column_to_print}{i}'].value)

In [4]:
column_printer(orders, 'A')

A1 Order_ID
A2 100000
A3 100001
A4 100002
A5 100003
A6 100004
A7 100005
A8 100006
A9 100007
A10 100008
A11 100009
A12 100010
A13 100011
A14 100012
A15 100013
A16 100014
A17 100020
A18 100021
A19 100022
A20 100023
A21 100024
A22 100025
A23 100026
A24 100027
A25 100031
A26 100032
A27 100033
A28 100034


In [5]:
column_printer(orders, 'D')

D1 Subtotal
D2 15.98
D3 899.97
D4 799.97
D5 117.96
D6 5.99
D7 599.99
D8 24.99
D9 1799.94
D10 99.99
D11 254.95
D12 29.98
D13 99.99
D14 25.98
D15 649.98
D16 89.99
D17 119.99
D18 599.99
D19 649.98
D20 24.99
D21 99.99
D22 99.99
D23 5.99
D24 24.99
D25 999.96
D26 99.99
D27 399.97
D28 89.99


In [6]:
column_printer(orders, 'A')

A1 Order_ID
A2 100000
A3 100001
A4 100002
A5 100003
A6 100004
A7 100005
A8 100006
A9 100007
A10 100008
A11 100009
A12 100010
A13 100011
A14 100012
A15 100013
A16 100014
A17 100020
A18 100021
A19 100022
A20 100023
A21 100024
A22 100025
A23 100026
A24 100027
A25 100031
A26 100032
A27 100033
A28 100034


In [7]:
column_printer(orders, 'B')

B1 Customer_ID
B2 C00004
B3 C00007
B4 C00015
B5 C00016
B6 C00020
B7 C00010
B8 C00006
B9 C00001
B10 C00003
B11 C00014
B12 C00001
B13 C00001
B14 C00005
B15 C00008
B16 C00013
B17 C00004
B18 C00017
B19 C00019
B20 C00002
B21 C00008
B22 C00021
B23 C00022
B24 C00006
B25 C00018
B26 C00018
B27 C00010
B28 C00016


In [8]:
column_printer(orders, 'C')

C1 Order_Date
C2 11/26/2021
C3 11/26/2021
C4 11/26/2021
C5 11/26/2021
C6 11/26/2021
C7 11/26/2021
C8 11/26/2021
C9 11/26/2021
C10 11/26/2021
C11 11/26/2021
C12 11/26/2021
C13 11/26/2021
C14 11/26/2021
C15 11/26/2021
C16 11/26/2021
C17 11/27/2021
C18 11/27/2021
C19 11/27/2021
C20 11/27/2021
C21 11/27/2021
C22 11/27/2021
C23 11/27/2021
C24 11/28/2021
C25 11/28/2021
C26 11/28/2021
C27 11/28/2021
C28 11/28/2021


In [9]:
column_printer(orders, 'D')

D1 Subtotal
D2 15.98
D3 899.97
D4 799.97
D5 117.96
D6 5.99
D7 599.99
D8 24.99
D9 1799.94
D10 99.99
D11 254.95
D12 29.98
D13 99.99
D14 25.98
D15 649.98
D16 89.99
D17 119.99
D18 599.99
D19 649.98
D20 24.99
D21 99.99
D22 99.99
D23 5.99
D24 24.99
D25 999.96
D26 99.99
D27 399.97
D28 89.99


In [10]:
column_printer(orders, 'E')

E1 Tax
E2 None
E3 None
E4 None
E5 None
E6 None
E7 None
E8 None
E9 None
E10 None
E11 None
E12 None
E13 None
E14 None
E15 None
E16 None
E17 None
E18 None
E19 None
E20 None
E21 None
E22 None
E23 None
E24 None
E25 None
E26 None
E27 None
E28 None


In [11]:
column_printer(orders, 'F')

F1 Total
F2 None
F3 None
F4 None
F5 None
F6 None
F7 None
F8 None
F9 None
F10 None
F11 None
F12 None
F13 None
F14 None
F15 None
F16 None
F17 None
F18 None
F19 None
F20 None
F21 None
F22 None
F23 None
F24 None
F25 None
F26 None
F27 None
F28 None


In [12]:
column_printer(orders, 'G')

G1 Location
G2 Sun Valley
G3 Stowe
G4 Mammoth
G5 Stowe
G6 Sun Valley
G7 Mammoth
G8 Mammoth
G9 Mammoth
G10 Sun Valley
G11 Sun Valley
G12 Mammoth
G13 Mammoth
G14 Sun Valley
G15 Stowe
G16 Sun Valley
G17 Sun Valley
G18 Stowe
G19 Sun Valley
G20 Stowe
G21 Stowe
G22 Mammoth
G23 Sun Valley
G24 Mammoth
G25 Stowe
G26 Stowe
G27 Mammoth
G28 Stowe


In [13]:
column_printer(orders, 'H')

H1 Items_Ordered
H2 10001, 10002
H3 10008, 10009, 10010
H4 10011, 10012, 10013
H5 10002, 10003, 10004, 10006
H6 10001
H7 10010
H8 10004
H9 10008, 10008, 10009, 10009, 10009, 10010, 10010
H10 10005
H11 10002, 10003, 10004, 10006, 10007
H12 10002, 10003
H13 10005
H14 10001, 10003
H15 10012, 10013
H16 10014
H17 10007
H18 10010
H19 10012, 10013
H20 10004
H21 10005
H22 10008
H23 10001
H24 10002
H25 10005, 10008, 10009, 10010
H26 10006
H27 10005, 10008, 10009
H28 10014


### Order Data Dictionary

Create a dictionary with all of the information contained in the 'Orders_Info' Worksheet.

* The dictionary keys should be Order IDs (Column A)
* The values should a list storing the data in the rest of the columns.
* The columns in the list stored as values should be B, C, D, G, and H (after converting to list). 
* For example: The first order, column H should be the nested list: [10001, 10002]


A few notes:

* Convert the Items_Ordered field into a list (we covered a helpful string method for this)
* This can be done with a dictionary comprehension
* DO NOT INCLUDE EXCEL COLUMN HEADERS

In [14]:
order_id = []

for i, cell in enumerate(orders['A'], start= 1):
    if i != 1:
        order_id.append(cell.value)

order_id
len(order_id)

27

In [15]:
customer_id = []

for i, cell in enumerate(orders['B'], start= 1):
    if i != 1:
        customer_id.append(cell.value)

customer_id
len(customer_id)

27

In [16]:
order_date = []

for i, cell in enumerate(orders['C'], start= 1):
    if i != 1:
        order_date.append(cell.value)

order_date
len(order_date)

27

In [17]:
subtotal = []

for i, cell in enumerate(orders['D'], start= 1):
    if i != 1:
        subtotal.append(cell.value)

subtotal
len(subtotal)

27

In [18]:
location = []

for i, cell in enumerate(orders['G'], start= 1):
    if i != 1:
        location.append(cell.value)

location
len(location)

27

In [19]:
items_ordered = []

for i, cell in enumerate(orders['H'], start= 1):
    if i != 1:
        items_ordered.append(cell.value)

items_ordered


['10001, 10002',
 '10008, 10009, 10010',
 '10011, 10012, 10013',
 '10002, 10003, 10004, 10006',
 10001,
 10010,
 10004,
 '10008, 10008, 10009, 10009, 10009, 10010, 10010',
 10005,
 '10002, 10003, 10004, 10006, 10007',
 '10002, 10003',
 10005,
 '10001, 10003',
 '10012, 10013',
 10014,
 10007,
 10010,
 '10012, 10013',
 10004,
 10005,
 10008,
 10001,
 10002,
 '10005, 10008, 10009, 10010',
 10006,
 '10005, 10008, 10009',
 10014]

In [20]:
order_dict = {
    orders[f'A{order}'].value:[
    orders[f'B{order}'].value,
    orders[f'C{order}'].value,
    orders[f'D{order}'].value,
    orders[f'G{order}'].value,
    str(orders[f'H{order}'].value).split(', ')
    ]
    for order in range(2, orders.max_row + 1)    
}

In [21]:
pprint(order_dict)

{100000: ['C00004', '11/26/2021', 15.98, 'Sun Valley', ['10001', '10002']],
 100001: ['C00007', '11/26/2021', 899.97, 'Stowe', ['10008', '10009', '10010']],
 100002: ['C00015',
          '11/26/2021',
          799.97,
          'Mammoth',
          ['10011', '10012', '10013']],
 100003: ['C00016',
          '11/26/2021',
          117.96,
          'Stowe',
          ['10002', '10003', '10004', '10006']],
 100004: ['C00020', '11/26/2021', 5.99, 'Sun Valley', ['10001']],
 100005: ['C00010', '11/26/2021', 599.99, 'Mammoth', ['10010']],
 100006: ['C00006', '11/26/2021', 24.99, 'Mammoth', ['10004']],
 100007: ['C00001',
          '11/26/2021',
          1799.94,
          'Mammoth',
          ['10008', '10008', '10009', '10009', '10009', '10010', '10010']],
 100008: ['C00003', '11/26/2021', 99.99, 'Sun Valley', ['10005']],
 100009: ['C00014',
          '11/26/2021',
          254.95,
          'Sun Valley',
          ['10002', '10003', '10004', '10006', '10007']],
 100010: ['C00001', '11/

### Sales Tax Calculation (Last time!)

We need to calculate the sales tax and total amount owed for every order in this sheet.

* If location is Sun Valley, apply a sales tax of 8%
* If location is Mammoth, apply a sales tax of 7.75%.
* If location is Stowe, apply a sales tax of 6%

Use the tax_calculator function to apply sales tax to each subtotal.

**Insert** the calculated sales tax and total amounts into your customer dictionary immediately after subtotal.

**Tip**: Figure out how to do this once and then loop through the column.

In [22]:
for order in order_dict.values():
    if order[3] == "Sun Valley":
        transaction = tc.tax_calculator(order[2], 0.08)
    elif order[3] == "Mammoth":
        transaction = tc.tax_calculator(order[2], 0.0775)
    else:
        transaction = tc.tax_calculator(order[2], 0.06)
    order.insert(3, transaction[1])
    order.insert(4,  transaction[2])

In [23]:
pprint(order_dict)

{100000: ['C00004',
          '11/26/2021',
          15.98,
          1.28,
          17.26,
          'Sun Valley',
          ['10001', '10002']],
 100001: ['C00007',
          '11/26/2021',
          899.97,
          54.0,
          953.97,
          'Stowe',
          ['10008', '10009', '10010']],
 100002: ['C00015',
          '11/26/2021',
          799.97,
          62.0,
          861.97,
          'Mammoth',
          ['10011', '10012', '10013']],
 100003: ['C00016',
          '11/26/2021',
          117.96,
          7.08,
          125.04,
          'Stowe',
          ['10002', '10003', '10004', '10006']],
 100004: ['C00020', '11/26/2021', 5.99, 0.48, 6.47, 'Sun Valley', ['10001']],
 100005: ['C00010', '11/26/2021', 599.99, 46.5, 646.49, 'Mammoth', ['10010']],
 100006: ['C00006', '11/26/2021', 24.99, 1.94, 26.93, 'Mammoth', ['10004']],
 100007: ['C00001',
          '11/26/2021',
          1799.94,
          139.5,
          1939.44,
          'Mammoth',
          ['10008', '

### Write Sales Tax and Total Into the Excel Sheet


Great job! Now we just need to write this data into the workbook.

Write the sales tax and total you just calculated into the workbook, then save!

Call this workbook 'maven_ski_shop_data_fixed'.

**Tip:** There are a few ways to do this. As always, be patient, solve one step at a time.

In [24]:
for i, dict_order in enumerate(order_dict.values(), start=2):
    orders[f'E{i}'] = dict_order[3]
    orders[f'F{i}'] = dict_order[4]

In [25]:
wb.save('maven_ski_shop_data_fixed.xlsx')

## Part 2

### Analysis Time!


Now that we've fixed the data - it's time to perform analysis on our sales.

Our starting point will be the order_dict we created, after we added the taxes and totals columns.

Our first step will be to write a function that calculates the sum of a 'column' of data in our dictionary.

A 'column' for example, would be subtotals, which is at index 2 in the list stored as our dictionary values.

Your function should take the following arguments:

* Column Index (the index of the data of interest in the list stored in our dictionary values)
* Dictionary Name

It should output:

* The sum of values in the column (rounded to two decimal places)

Assume only numeric values will be in the column (You can develop cleaning logic later :D)

**Tip:** Use a list comprehension to retrieve the values of interest.

In [26]:
order_dict[100033]

['C00010',
 '11/28/2021',
 399.97,
 31.0,
 430.97,
 'Mammoth',
 ['10005', '10008', '10009']]

In [27]:
def column_sum(column_index, dictionary):
    return round(sum(value[column_index] for value in dictionary.values()), 2)


#### Sum The Subtotal, Tax, and Total Columns 

Now that we have our column sum function, calculate the sum of:

* Subtotals
* Taxes
* Totals

In [28]:
print(column_sum(2, order_dict))
print(column_sum(3, order_dict))
print(column_sum(4, order_dict))

8731.47
617.2
9348.67


### What is the average of our subtotals?

Calculate the average value of our transactions.

Remember that each entry in your dictionary is one order.

In [29]:
round(column_sum(2, order_dict) / len(order_dict), 2)

323.39

### How many unique customers did we have?

Calculate the total number of unique customers in our sales data.

Then calculate the number of orders per customer (total orders / unique customers)

In [30]:
number_of_unique_customers = len({value[0] for value in order_dict.values()})
number_of_unique_customers

19

In [31]:
number_of_orders_per_customer = round(len(order_dict) / number_of_unique_customers, 1)
number_of_orders_per_customer

1.4

### How many items in total did we sell?

Calculate the total number of items we sold in across all orders. 

This information is in Column H, which should be the last element in order_dict's values.

In [32]:
sum(len(value[6]) for value in order_dict.values())

54

### Sales By Location

Calculate the sum of subtotals by location.

Create a dictionary to store them, where location is the key, and revenue for that location is the value.

A few steps to consider:

* Loop through your dictionary
* build a dictionary as you go with location as key
* increment revenue every time a transaction matches the location.


Your output should look like 
{'Location1': sum of subtotals for 'Location1'}

With an entry for each location.

In [33]:
location_sums = {}

for value in order_dict.values():
    location = value[5]
    if location not in location_sums:
        location_sums[location] = 0
    else:
        location_sums[location] += value[2]

location_sums

{'Sun Valley': 1252.86,
 'Stowe': 2682.8499999999995,
 'Mammoth': 3079.8399999999992}

In [34]:
location_sums = {}

for value in order_dict.values():
    location = value[5]
    location_sums[location] = round(location_sums.get(location,0) + value[2], 2)
location_sums

{'Sun Valley': 1268.84, 'Stowe': 3582.82, 'Mammoth': 3879.81}

### Challenge: Aggregator Function

Now that you've summed revenue by category (you just summed by location), can you write a function to generalize calculating a sum of a column, grouped by the unique values in another column? (for example, sum of totals by date or customer_id).

Your function should take the following arguments:
* index of the 'column' (index position in order_dict) to group by
* index of the 'column' (index position in order_dict) to sum by category
* the dictionary where the data is located (assume the same structure as order_dict.

It should return:
* A dictionary with the categories as keys, and the sum by category as value.

Once you've done so use your function to sum totals by date and customer_id.

In [35]:
def aggregator(category_index, col_to_sum_index, dictionary):
    
    category_sums = {}   
    
    for value in dictionary.values():
        category = value[category_index]
        category_sums[category] = round(category_sums.get(category,0) + value[col_to_sum_index], 2)

    return category_sums

In [36]:
# index 1 = date, index 4 = total
aggregator(1, 4, order_dict)

{'11/26/2021': 5915.18, '11/27/2021': 1714.25, '11/28/2021': 1719.24}

In [37]:
# index 0 = customer_id
aggregator(0, 4, order_dict)

{'C00004': 146.85,
 'C00007': 953.97,
 'C00015': 861.97,
 'C00016': 220.43,
 'C00020': 6.47,
 'C00010': 1077.46,
 'C00006': 53.86,
 'C00001': 2079.48,
 'C00003': 107.99,
 'C00014': 275.35,
 'C00005': 28.06,
 'C00008': 794.97,
 'C00013': 97.19,
 'C00017': 635.99,
 'C00019': 701.98,
 'C00002': 26.49,
 'C00021': 107.74,
 'C00022': 6.47,
 'C00018': 1165.95}